compiler/rustc_ast_passes/src/ast_validation.rs

```diff
@@ -213,6 +229,85 @@ impl<'a> AstValidator<'a> {
+                if let Some(Ident { name: kw::Underscore, .. }) = ident {
+                    // Empty because we want to error only if Ident != `_`
+                } else {
+                    let struct_kind = if matches!(ty.kind, TyKind::AnonymousStruct(..)) {
+                        "struct"
+                    } else {
+                        "union"
+                    };
+                    let msg =
+                        format!("anonymous {} not allowed with a named identifier", struct_kind);
+                    self.err_handler()
+                        .struct_span_err(field.span, msg.as_str())
+                        // There are no `None` idents on a struct field def, so it should be safe to unwrap
+                        .span_suggestion(
+                            ident.unwrap().span,
+                            "replace with",
+                            "_".to_string(),
+                            Applicability::MachineApplicable,
+                        )
+                        .emit();
+                }
+            }
+            TyKind::Path(..) => {
+                // If the anonymous field contains a Path as type, we can't determine if the path is a valid
+                // struct or union, so we skip the check
+            }
+            _ => {
+                if let Some(Ident { name: kw::Underscore, .. }) = ident {
+                    let msg = "only anonymous structs and unions are allowed";
+                    let label = "not an anonymous struct or union";
+                    self.err_handler()
+                        .struct_span_err(field.span, msg)
+                        .span_label(ty.span, label)
+                        .emit();
+                }
+            }
+        }
+    }
+
+    fn deny_anonymous_struct(&self, ty: &Ty) {
+        match &ty.kind {
+            TyKind::AnonymousStruct(..) => {
+                self.err_handler()
+                    .struct_span_err(
+                        ty.span,
+                        "anonymous struct not allowed outside of a field struct or union",
+                    )
+                    .span_label(ty.span, "anonymous struct declared here")
+                    .emit();
+            }
+            TyKind::AnonymousUnion(..) => {
+                self.err_handler()
+                    .struct_span_err(
+                        ty.span,
+                        "anonymous union not allowed outside of a field struct or union",
+                    )
+                    .span_label(ty.span, "anonymous union declared here")
+                    .emit();
+            }
+            _ => {}
+        }
+    }
+
+    fn deny_anonymous_field(&self, field: &FieldDef) {
+        if let Some(Ident { name: kw::Underscore, span }) = field.ident {
+            self.err_handler()
+                .struct_span_err(
+                    field.span,
+                    "anonymous field not allowed outside of a field struct or union",
```

Comment:
```suggestion
                    "anonymous fields are not allowed outside of structs or unions",
```
(Full sentences in primary messages, abbreviated sentences in labels.)